torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -629,6 +644,74 @@ void RequestCallbackImpl::processRpc(
       markComplete(std::move(CleanupAutogradContextResp()).toMessage());
       return;
     }
+    case MessageType::RUN_WITH_PROFILING_REQ: {
+      auto& rpcWithProfilingReq = static_cast<RpcWithProfilingReq&>(rpc);
+      auto wrappedMsgType = rpcWithProfilingReq.wrappedMessageType();
+      const auto profilingConfig = rpcWithProfilingReq.getProfilingConfig();
+      auto wrappedRpcResponseFuture = std::make_shared<FutureMessage>();
+      auto fromWorkerId = rpcWithProfilingReq.fromWorkerId();
+      // Enable the profiler with the config from the sender.
+      std::vector<torch::autograd::profiler::Event> profiledEvents;
+      {
+        torch::autograd::profiler::TLSProfilerGuard g(
+            profilingConfig,
+            [&profiledEvents](const std::vector<
+                              std::vector<torch::autograd::profiler::Event>>&
+                                  event_lists) {
+              // Gather all events into a vector
+              for (auto& l : event_lists) {
+                for (auto& e : l) {
+                  profiledEvents.push_back(e);
+                }
+              }
+              // find __start_profile event.
+              const torch::autograd::profiler::Event* profilerStart = nullptr;
+              for (const auto& e : profiledEvents) {
+                if (0 == strcmp(e.name(), "__start_profile")) {
+                  profilerStart = &e;
+                  break;
+                }
+              }
+              // We should always find __start_profile.
+              TORCH_CHECK(profilerStart != nullptr);
+              // For every event, precompute the relative CPU time before
+              // serialization.
+              for (const auto& e : profiledEvents) {
+                double cpu_elapsed_us = profilerStart->cpu_elapsed_us(e);
+                e.setRelativeElapsedCpuUs(cpu_elapsed_us);
+              }
+            });
+        TORCH_INTERNAL_ASSERT(
+            torch::autograd::profiler::profilerEnabled(),
+            "Expected profiler to be enabled!");
+        // Kick off processing for nested work and get Future<T> result in
+        // wrappedRpcResponseFuture
+        processRpc(
+            rpcWithProfilingReq.wrappedRpc(),
+            wrappedMsgType,
+            messageId,
+            wrappedRpcResponseFuture);
+      }
+      wrappedRpcResponseFuture->addCallback([wrappedRpcResponseFuture,
+                                             responseFuture,
+                                             fromWorkerId,
+                                             profiledEvents =
+                                                 std::move(profiledEvents)] {
+        if (wrappedRpcResponseFuture->hasError()) {
+          // Propagate error
+          responseFuture->setError(wrappedRpcResponseFuture->error()->what());
```

Comment:
This means error RPC requests are not profiled?

Reply:
Yes, no plan to profile these. We don't really have a pattern of wrapping errors with a successful response message which is what we'd want to do here.